### DISCUSSION TITLE
Renovate not committing go.mod with updates in some cases

### DISCUSSION THREAD

**Discussion by chancez** [renovatebot/renovate#30358] [Request Help]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

37.440.7

### Please tell us more about your question or problem

So I've contribute a little a bit on the gomod updater previously and I suspect I see where the problem is, but basically I'm seeing some cases where renovate runs `go work vendor` and `go work sync` and there are certain changes to `go.mod` that are not being committed in the PRs.

I can see it reads the latest `go.mod` after all the other commands run : https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/gomod/artifacts.ts#L400-L427, which is what cause  so it should be seeing the latest updates to the file too.

To be clear, Renovate is creating a PR with updates to our `go.mod`, but it's missing 1 line of change that removes an entry from the `go.mod`. 

For example, I can see the logs detecting updates to `go.mod`:

```
2024-07-24T18:11:30.0716391Z DEBUG: Returning updated go.sum (repository=isovalent/hubble-timescape, baseBranch=v1.4, branch=renovate/v1.4-patch-all-go-module-dependencies-v1.4)
2024-07-24T18:11:30.0721138Z DEBUG: Returning updated go.work.sum (repository=isovalent/hubble-timescape, baseBranch=v1.4, branch=renovate/v1.4-patch-all-go-module-dependencies-v1.4)
2024-07-24T18:11:30.1033487Z DEBUG: Found updated go.mod after go.sum update (repository=isovalent/hubble-timescape, baseBranch=v1.4, branch=renovate/v1.4-patch-all-go-module-dependencies-v1.4)
```

But we have CI which runs `go work vendor` to verify someone didn't manually modify our vendor directory without updating the `go.mod` as well, and it's getting a diff. 

I looked at the recent changes to the `gomod` manager: https://github.com/renovatebot/renovate/commits/main/lib/modules/manager/gomod and there's only 3 changes since my most recent PR, so if it's a regression, it would be one of those commits. 

- The [first commit](https://github.com/renovatebot/renovate/commit/d62444b97dd5bfb87c02e4a11cb83d11c660742f) is a simple addition of a newline, so that probably isn't it
- The [second commit](https://github.com/renovatebot/renovate/commit/5f21e173982fde6c30c9df33b2c4713a16e21c59) adds an option to skip vendoring, and we don't set that option, and I can see from our logs it's still running all the usual vendoring steps.
- The [third commit](https://github.com/renovatebot/renovate/commit/50191defcc4666d112bf5939464dc28ed52fc9bf) changed a bit more, but it _should_ just be a variable name change from what I can tell, so I don't think it's that either.

So I can only see this being a bug that already existed that we just haven't hit yet for some reason.

While writing this out, I did come up with another potential reason it's not working correctly:  Perhaps it's an ordering issue. We have two `go.mod` files. We have `./api/go.mod` and `./go.mod` and a root `./go.work`. 

I can see it's updating the root `go.mod` first, which might be the problem. If it's due to `go work sync` then it might depend on the updates to `api/go.mod` and running `go work sync` after `api/go.mod` is syncing to the root `go.mod`, after the changes have already been staged.

```
2024-07-24T18:10:11.0221412Z DEBUG: gomod.updateArtifacts(go.mod) (repository=isovalent/hubble-timescape, baseBranch=v1.4, branch=renovate/v1.4-patch-all-go-module-dependencies-v1.4)
...
2024-07-24T18:11:30.1044895Z DEBUG: gomod.updateArtifacts(api/go.mod) (repository=isovalent/hubble-timescape, baseBranch=v1.4, branch=renovate/v1.4-patch-all-go-module-dependencies-v1.4)
```

However, this makes things a bit more complicated since I'm not sure how to fix that. We do have a `postUpgradeTask` that runs vendoring for consistency checks, but it doesnt' seem to be helping:
```
2024-07-24T18:11:44.5106726Z DEBUG: Executed post-upgrade task (repository=isovalent/hubble-timescape, baseBranch=v1.4, branch=renovate/v1.4-patch-all-go-module-dependencies-v1.4)
2024-07-24T18:11:44.5109410Z        "dep": "github.com/isovalent/hubble-k8s-exporter github.com/isovalent/hubble-k8s-exporter",
2024-07-24T18:11:44.5110867Z        "cmd": "make renovate-vendor",
```

This post upgrade task basically executes:
```
	$(GO) mod tidy
	$(GO) work sync
	$(GO) work vendor
	$(GO) mod verify
```

So it should be updating the root `go.mod`, especially since it runs after the previous ` gomod.updateArtifacts` calls. If I run it locally I get the same diff our CI is seeing to the `go.mod` that renovate isn't committing. 

Renovate also logs the following after running the `postUpgradeTask`:
```
2024-07-24T18:12:55.5470757Z DEBUG: Post-upgrade file saved (repository=isovalent/hubble-timescape, baseBranch=v1.4, branch=renovate/v1.4-patch-all-go-module-dependencies-v1.4)
2024-07-24T18:12:55.5475558Z        "dep": "github.com/isovalent/hubble-k8s-exporter github.com/isovalent/hubble-k8s-exporter",
2024-07-24T18:12:55.5476579Z        "file": "go.mod",
2024-07-24T18:12:55.5477089Z        "pattern": "**/*"
```

So im not sure why it's not getting the update to the `go.mod` with the line being removed.

Is there a way we could configure the ordering between two separate `gomod` managers running so it does updates on `api/go.mod` first, then updates the root `go.mod` afterwards?

Any ideas?

**Comment by chancez** [comment]
I see some stuff about `fileReplacePosition` in the project related to fixing some Maven updates (https://github.com/renovatebot/renovate/pull/8263/) and I wonder if this is something we could leverage to fix this problem.

**Reply by chancez** [reply]
Seems like switching the `gomod` manager to using `extractAllPackageFiles` instead of `extractPackageFile` could be a resolution for this based on the `adding-a-package-manager.md` docs.

**Reply by chancez** [reply]
I see roughly what needs to be done I think, however I am not entirely sure how to actually do the ordering. I know I want `api/go.mod` to be updated first, then `go.mod`. However, to know that I would have to see that the `api` module is listed in `go.mod`.

So roughly:
- Get all the deps for each `go.mod` (packageFile). Also store the `module` from the `go.mod` since we'll need it later.
- If a `go.work` exists, look at the `go.mod` files listed in the directories specified in the `use`.
- For each `go.mod`, if it has a `depName` that is equal to a the `module` of another `go.mod` listed in the `go.work`, then it needs to be updated after it's dependent `go.mod`

I think?

**Reply by rarkins** [reply]
This sorting/reordering could perhaps be done without needing to change to `extractAllPackageFiles`

**Comment by chancez** [comment]
As much as I want to fix this, it might take some time. I'd love a short term solution that ensures the `go.mod` is added to the PR after we run our `postUpgradeTask`.

**Reply by rarkins** [reply]
Could that be done with a dummy postUpgradeTasks (e.g. `ls`) plus `fileFilters` including `go.mod`?

**Reply by chancez** [reply]
🤔 I could try it!

**Reply by chancez** [reply]
Actually, I'm not sure this will do anything. We don't set `fileFilters`, and you can only have one `postUpgradeTasks` (but with many commands).  For reference, this is what ours looks like:
```
    {
      "groupName": "all go module dependencies {{ baseBranch }}",
      "matchFiles": [
        "go.work",
        "go.mod",
        "go.sum",
        "api/go.mod",
        "api/go.sum",
      ],
      "matchUpdateTypes": [
        "major",
        "minor",
        "digest",
        "patch",
        "pin",
        "pinDigest"
      ],
      "postUpgradeTasks": {
        "commands": [
          // Install Go inside renovate so we can run `make renovate-vendor`
          "make renovate-install-go",
          "make renovate-vendor",
          // when golang deps are updated, make sure we keep licenses in sync
          "make licenses.csv",
        ],
        "executionMode": "branch",
      },
      "addLabels": [
        "area/vendor",
        "release-note/misc"
      ]
    },
```

So I'm not really sure how we can apply your suggestion and expect it to change anything.

**Reply by chancez** [reply]
Well, as a work-around, I disabled vendoring (using the new `gomodSkipVendor` option), and opted to just have vendoring 100% handled by our postUpgradeTasks.

What I find funny is this is the option I had originally proposed to work around our vendoring problems, but someone else implemented because I figured "hey I'll just make Renovate handle Go workspaces instead" 😂 . 

I'm not really sure what renovate does differently that causes the discrepancy. When I switch to our `make renovate-vendor` it's not removing the line anymore, so I'm not sure why it was being removed, but not committed when renovate handled vendoring, despite it showing in the diff when running our own `make renovate-vendor` command 😕 .
